From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Dreeam <61569423+Dreeam-qwq@users.noreply.github.com>
Date: Wed, 12 Oct 2022 10:42:15 -0400
Subject: [PATCH] Leaf config

Leaf Config v3
including load config, backup old or outdated config, and add config to spark profiler automatically.

TODO - Dreeam:
Add per world config
Update config reload:
    - Use different state to separate different configs reload
    (Or sync reload)

diff --git a/net/minecraft/server/Main.java b/net/minecraft/server/Main.java
index 31ce71305260f082729780d4e8fb4dba557f9883..170d2bac2c280921248c86d266add46260c5a76d 100644
--- a/net/minecraft/server/Main.java
+++ b/net/minecraft/server/Main.java
@@ -100,6 +100,8 @@ public class Main {
                 writePidFile(path);
             }
 
+            org.dreeam.leaf.config.LeafConfig.loadConfig(); // Leaf - Leaf config
+
             CrashReport.preload();
             if (optionSet.has("jfrProfile")) { // CraftBukkit
                 JvmProfiler.INSTANCE.start(Environment.SERVER);
@@ -116,6 +118,7 @@ public class Main {
             Bootstrap.bootStrap();
             Bootstrap.validate();
             Util.startTimerHackThread();
+            org.dreeam.leaf.config.ConfigModules.loadAfterBootstrap(); // Leaf - Leaf config - post load
             Path path1 = Paths.get("server.properties");
             DedicatedServerSettings dedicatedServerSettings = new DedicatedServerSettings(optionSet); // CraftBukkit - CLI argument support
             dedicatedServerSettings.forceSave();
diff --git a/net/minecraft/server/MinecraftServer.java b/net/minecraft/server/MinecraftServer.java
index b2af6755d55e37eb5cdcb0f78ffcb09854db676f..bc7f2a98442995413fb941c10b13d5d4e1f778ea 100644
--- a/net/minecraft/server/MinecraftServer.java
+++ b/net/minecraft/server/MinecraftServer.java
@@ -1290,6 +1290,7 @@ public abstract class MinecraftServer extends ReentrantBlockableEventLoop<TickTa
             org.spigotmc.WatchdogThread.tick();
             // Paper end
             org.spigotmc.WatchdogThread.hasStarted = true; // Paper
+            org.dreeam.leaf.config.LeafConfig.regSparkExtraConfig(); // Leaf - Leaf config
             // Paper start - Add onboarding message for initial server start
             if (io.papermc.paper.configuration.GlobalConfiguration.isFirstStart) {
                 LOGGER.info("*************************************************************************************");
